### ISSUE TITLE
Pillow 5.1.0's included dylibs binaries stopped working on Mac El Capitan (10.11)

### ISSUE THREAD

**Issue by varenc** [python-pillow/Pillow#3068]
### What did you do?
Upgraded pillow from 5.0 to 5.1

### What did you expect to happen?
For Pillow to keep working, or at least for the Pillow 5.1.0 change log to report that only 10.12+ is supported.

### What actually happened?
When trying to use pillow through the python `captcha` package, I got this error:
```
  File ".../venv/lib/python3.6/site-packages/captcha/image.py", line 11, in <module>
    from PIL import Image
ImportError: .../venv/lib/python3.6/site-packages/PIL/_imaging.cpython-36m-darwin.so, 2): Symbol not found: _clock_gettime
Referenced from: .../venv/lib/python3.6/site-packages/PIL/.dylibs/liblzma.5.dylib (which was built for Mac OS X 10.12)
  Expected in: /usr/lib/libSystem.B.dylib
 in .../venv/lib/python3.6/site-packages/PIL/.dylibs/liblzma.5.dylib

```
### What versions of Pillow and Python are you using?
Pillow 5.1.0 has the problem.  Running Python 3.6.5.

To reproduce this problem try testing Pillow on a 10.11 system.  Specifically, just import PIL.image to trigger this.

Installing `pillow==5.0.0` fixed the issue.

Did you intentionally drop 10.11 support?  If not it'd be great for this to keep working! Probably just means rebuilding the dylibs with 10.11 support.

If you meant to drop support than just clearer messaging around this would be helpful. (changelog, explicit version check and error message)

**Comment by wiredfool** [comment]
That’s unintentional.  

@matthew-brett ?  @radarhere ?

**Comment by wiredfool** [comment]
That does appear to fit the problem.

**Comment by varenc** [comment]
I'm no expert in this stuff, but it sounds like xcode9.2 simply defaults to building for 10.12, but you can still tell it to target older versions if you want.

I think you can do this by setting the `MACOSX_DEPLOYMENT_TARGET=10.11` env which xcode will pick up. I made a pull request on my fork of pillow-wheels that shows this, though can't actually invest the time to test this right now: https://github.com/varenc/pillow-wheels/pull/1  :-)

**Comment by wiredfool** [comment]
I’ve pushed a branch with the old Xcode, builds succeeded, so I think we’re back to a full OS X build. 

Look here http://a365fff413fe338398b6-1c8a9b3114517dc5fe17b7c3f8c63a43.r19.cf2.rackcdn.com/ for pillow 5.1.0 builds, anything done on 4/4 is my test, and should work as far back as the previous release.

**Comment by varenc** [comment]
I can confirm [this new .whl](http://a365fff413fe338398b6-1c8a9b3114517dc5fe17b7c3f8c63a43.r19.cf2.rackcdn.com/Pillow-5.1.0-cp36-cp36m-macosx_10_6_intel.macosx_10_9_intel.macosx_10_9_x86_64.macosx_10_10_intel.macosx_10_10_x86_64.whl) has me up and running with pillow 5.1.0 now!

looking at `liblzma.5.dylib` with otool now shows that the `LC_VERSION_MIN_MACOSX` is `10.10`.  If you do want to take advantage of xcode 9 and maybe support even older versions, tweaking `MACOSX_DEPLOYMENT_TARGET` might be the way to go, but happy I got things working now. Thanks!

**Comment by wiredfool** [comment]
Ok, good to hear that it’s working.  I think we’ll just stick with Xcode 6.4 for now, and do a 5.1.1 release shortly to fix the binaries.

**Comment by matthew-brett** [comment]
How about releasing a binary build with a build tag?

https://github.com/MacPython/wiki/wiki/Build-Tags

It's just a question of building the wheels, downloading and renaming (see script on that page) and then uploading.  Oh - and deleting the previous offending wheels.

**Comment by matthew-brett** [comment]
I'm happy to help with the binary build ...

**Comment by jbarlow83** [comment]
I don't think it's a good idea to replace a known-bad version after it's been out for a while. Just bump the version.

People will be fixing their builds by blacklisting the bad release:

```
install_requires=['Pillow != 5.1.0 ; platform_system = "Darwin"']
```

**Comment by varenc** [comment]
my two cents:  If using build-tags lets you get the fixed 5.1.0 binaries out sooner then that seems preferable. But if 5.1.1 will get released soon anyway, then just go with that.

**Comment by WoLpH** [comment]
Or do both, replace the old binary and push a new release do it works for
everyone.

**Comment by matthew-brett** [comment]
I can't see the downside to fixing the binaries.   OK some people may be skipping the release, but they won't be further hurt by fixed binaries.  If the bad version did anything special other than crash I'd agree, but replacing crash by no-crash seems like an easy win.

**Comment by hugovk** [comment]
Please see https://github.com/python-pillow/Pillow/issues/3089.

**Comment by hugovk** [comment]
What's the best way to get pillow-wheels building on Xcode 8?

https://github.com/python-pillow/pillow-wheels/pull/79 is the PR that upgraded from 8 to 9.2, because the latest multibuild was missing webp and upgrading fixed it.
 
Upgrading to the latest multibuild was needed because rsync was missing (with but perhaps not limited to 8): https://github.com/python-pillow/pillow-wheels/issues/76.

**Comment by hugovk** [comment]
To answer my own question, from upthread:

> I’ve pushed a branch with the old Xcode, builds succeeded, so I think we’re back to a full OS X build.

Here's a PR, it's with Xcode 6.4: https://github.com/python-pillow/pillow-wheels/pull/84.

**Comment by radarhere** [comment]
Does anyone have any thoughts on what versions of macOS should be supported by pillow-wheels? We're talking about 10.11 here, so n-2? Or is it the case that we should just support as far back as we can without difficulty?

**Comment by WoLpH** [comment]
I would personally say 10.10 and above but if it's easy to support I would go for all available.

If osx wouldn't have died on me a few years back I'd still be running 10.10. Newer versions have always broken more than they've fixed for me

**Comment by hugovk** [comment]
I'd go for at least those supported by Apple, and further if there's no difference or not too much burden. 

Let's go with Xcode 6.4 now, because it's the easiest to fix with that :) Xcode 6.4 should support at least as far back as 10.10.4 Yosemite.

---

Wikipedia gives these support statuses:

* [OS X 10.10 Yosemite](https://en.wikipedia.org/wiki/OS_X_Yosemite): Unsupported as of September 2017, iTunes support to end in August 2018, similar to Mavericks.
* [OS X 10.11 El Capitan](https://en.wikipedia.org/wiki/OS_X_El_Capitan): Security updates and printer drivers only. Extended support to end this year, and iTunes support ends in next year.
* [macOS 10.12 Sierra](https://en.wikipedia.org/wiki/MacOS_Sierra): Extended support ends in September 2019 . iTunes, in August 2020.

---

Checking pip download stats for the last 30 days, very little for pre-10.10:

| system_name        | distro_version                                  | percent | download_count |
| ------------------ | ----------------------------------------------: | ------: | -------------: |
| Darwin             |                                         10.13.3 |   2.07% |         23,200 |
| Darwin             |                                         10.12.6 |   0.85% |          9,528 |
| Darwin             |                                         10.13.4 |   0.60% |          6,743 |
| Darwin             |                                         10.11.6 |   0.47% |          5,293 |
| Darwin             |                                         10.13.2 |   0.24% |          2,707 |
| Darwin             |                                         10.13.1 |   0.15% |          1,695 |
| Darwin             |                                         10.10.5 |   0.14% |          1,527 |
| Darwin             |                                         10.11.5 |   0.06% |            674 |
| Darwin             |                                         10.12.5 |   0.05% |            508 |
| Darwin             |                                           10.10 |   0.04% |            499 |
| Darwin             |                                           10.13 |   0.04% |            479 |
| Darwin             |                                            None |   0.03% |            364 |
| Darwin             |                                         10.12.4 |   0.03% |            296 |
| Darwin             |                                         10.12.3 |   0.02% |            264 |
| Darwin             |                                           10.12 |   0.02% |            230 |
| Darwin             |                                          10.9.5 |   0.02% |            178 |
| Darwin             |                                         10.12.1 |   0.01% |            136 |
| Darwin             |                                         10.11.4 |   0.01% |            125 |
| Darwin             |                                         10.12.2 |   0.01% |            107 |
| Darwin             |                                         10.13.5 |   0.01% |             80 |
| Darwin             |                                         10.11.3 |   0.01% |             59 |
| Darwin             |                                         10.11.2 |   0.00% |             44 |
| Darwin             |                                         10.11.1 |   0.00% |             40 |
| Darwin             |                                         10.10.3 |   0.00% |             33 |
| Darwin             |                                           10.11 |   0.00% |             33 |
| Darwin             |                                           10.14 |   0.00% |             29 |
| Darwin             |                                         10.13.6 |   0.00% |             23 |
| Darwin             |                                          10.7.5 |   0.00% |             18 |
| Darwin             |                                         10.10.4 |   0.00% |             17 |
| Darwin             |                                          10.8.5 |   0.00% |             14 |
| Darwin             |                                         10.10.2 |   0.00% |             14 |
| Darwin             |                                         10.10.1 |   0.00% |              9 |
| Darwin             |                                          10.6.8 |   0.00% |              8 |
| Darwin             |                                           6.1.4 |   0.00% |              7 |
| Darwin             |                                           10.15 |   0.00% |              2 |
| Darwin             |                                          10.9.4 |   0.00% |              1 |
| Darwin             |                                           10.90 |   0.00% |              1 |
| Darwin             |                                          10.3.1 |   0.00% |              1 |
| Darwin             |                                             9.1 |   0.00% |              1 |
| Darwin             |                                          10.9.2 |   0.00% |              1 |
| Darwin             |                                            10.9 |   0.00% |              1 |
| Darwin             |                                            10.8 |   0.00% |              1 |

(via `pypinfo --limit 1000 --percent --pip --markdown pillow system distro-version | grep Darwin`)

**Comment by wiredfool** [comment]
We're advertising support for 10.6 through 10.10, e.g.: Pillow-5.1.0-cp36-cp36m-macosx_10_6_intel.macosx_10_9_intel.macosx_10_9_x86_64.macosx_10_10_intel.macosx_10_10_x86_64.whl

Or, perhaps there's something _subtle_ in the naming convention that I'm not getting.  But it appears that the contents of the package don't match the naming. 

I'd say that in general, we should support currently shipping systems in binary, and the long tail in source. For anything but windows, pillow isn't that hard to install from source for someone who's dedicated enough to be using an out of support operating system. Either that or they can step up with some cash or labor.

**Comment by matthew-brett** [comment]
The naming pattern's a bit confusing I agree.  It dates from the time when pip running on a Python naming itself 10.8 didn't accept a wheel with name > 10.8.  That's long past now, so the name should just reflect the earliest OSX the wheel supports.  What is that - on XCode 6.4?

I suggest using the earliest XCode available unless there's a good reason not to.

**Comment by WoLpH** [comment]
Here's the version overview by major version (the minor ones aren't too relevant here I think):

Version | Count | Percent
--------|-------|--------
10.03 | 1 | 0.00%
10.06 | 8 | 0.01%
10.07 | 18 | 0.03%
10.08 | 15 | 0.03%
10.09 | 182 | 0.33%
10.10 | 2099 | 3.84%
10.11 | 6268 | 11.48%
10.12 | 11069 | 20.27%
10.13 | 34927 | 63.95%
10.14 | 29 | 0.05%
10.15 | 2 | 0.00%

**Comment by hugovk** [comment]
Xcode 6.4 is the earliest available on Travis CI:

https://docs.travis-ci.com/user/reference/osx/#Xcode-6.4

> What is that - on XCode 6.4?

I don't understand, please could you rephrase?

**Comment by wiredfool** [comment]
It occurs to me that if we're just renaming here to accurately reflect the supported platforms, we can just blow away all the OSX wheels and replace them with correctly named versions on pypi without a 5.1.1 release, and no need for build tags.

**Comment by hugovk** [comment]
Keeping 5.1.0 won't help with things like https://github.com/python-pillow/Pillow/issues/3068#issuecomment-380295263, but whatever's easier/quicker. It's never too long until the next quarterly release.

**Comment by wiredfool** [comment]
ok.

**Comment by denis-bz** [comment]
After `pip install Pillow` 9.4.0, matplotlib -> PIL -> 
cannot load 'libXdmcp.6.dylib' (load       command 0x80000034 is unknown)
Reinstall Pillow==9.3.0 works.

```
# matplotlib 3.6.2 -> PIL 9.4.0 on macos 10.14.6  2023-01-02
# https://github.com/python-pillow/Pillow/issues/3068 ?

import sys
import platform
import PIL

print( "mac_ver :", platform.mac_ver() )
print( "python sys.version :", sys.version )
print( "PIL.__version__ :", PIL.__version__ )

print( "\nimport matplotlib" )
import matplotlib
```

-->
```
mac_ver : ('10.14.6', ('', '', ''), 'x86_64')
python sys.version : 3.9.12 (v3.9.12:b28265d7e6, Mar 23 2022, 18:17:11)
[Clang 6.0 (clang-600.0.57)]
PIL.__version__ : 9.4.0

import matplotlib
Traceback (most recent call last):
  File "/Users/bz/etc/mac/mpl-pil-dylib.py", line 14, in <module>
    import matplotlib
  File "/opt/local/py/site/top/matplotlib/__init__.py", line 113, in <module>
    from . import _api, _version, cbook, _docstring, rcsetup
  File "/opt/local/py/site/top/matplotlib/rcsetup.py", line 27, in <module>
    from matplotlib.colors import Colormap, is_color_like
  File "/opt/local/py/site/top/matplotlib/colors.py", line 51, in <module>
    from PIL import Image
  File "/opt/local/py/site/top/PIL/Image.py", line 103, in <module>
    from . import _imaging as core
ImportError: dlopen(/opt/local/py/site/top/PIL/_imaging.cpython-39-darwin.so, 2): Library not       loaded: @loader_path/libXdmcp.6.dylib
  Referenced from: /opt/local/py/site/top/PIL/.dylibs/libxcb.1.1.0.dylib
  Reason: no suitable image found.  Did find:
    /opt/local/py/site/top/PIL/.dylibs/libXdmcp.6.dylib: cannot load 'libXdmcp.6.dylib' (load       command 0x80000034 is unknown)
    /opt/local/py/site/top/PIL/.dylibs/libXdmcp.6.dylib: cannot load 'libXdmcp.6.dylib' (load       command 0x80000034 is unknown)
```

Xcode, otool --
```
xcode-select version 2354.

 54224 Jan  2 10:59 libXdmcp.6.dylib

otool -L libXdmcp.6.dylib
libXdmcp.6.dylib:
	/DLC/PIL/.dylibs/libXdmcp.6.dylib (compatibility version 7.0.0, current version 7.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1311.100.3)
```

**Comment by radarhere** [comment]
> Is there any way of telling pip "don't use that wheel for 10.14" ?

Since you already know how to pin to the previous version's wheel, I don't know what you're after. You'd like to be able to run the same command on macOS 10.14 and install Pillow 9.3, and on later macOS versions, install the latest Pillow?

**Comment by denis-bz** [comment]
@radarhere, yes, exactly that: install the latest Pillow that works on my macOS.
"Pin to previous version": yes that's a fix
but after a `pip install Pillow` that looked ok: imho, an unforced error.

**Comment by radarhere** [comment]
I'm guessing that the change was because https://github.com/python-pillow/pillow-wheels/pull/350 updated libXau to 1.0.11. Homebrew removed support for 10.15 when making that update - https://github.com/Homebrew/homebrew-core/commit/712726ca875be4f7528c5288c48214ec7078ae99

I can't even test macOS 10.14 on GitHub Actions to investigate further - https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources

If you want to confirm my theory, you can test the wheels from https://github.com/radarhere/pillow-wheels/actions/runs/3825868768 - https://github.com/radarhere/pillow-wheels/suites/10148738168/artifacts/495813662. They are Pillow 9.4.0, but with libXau 1.0.10.

I don't think there's a way to adjust our wheels to work on macOS 10.14 without either removing or downgrading the dependency. You might be interested in building Pillow from source instead - `pip install Pillow --no-binary :all:`

**Comment by radarhere** [comment]
More specifically, you can `python3 -m pip install Pillow-9.4.0-cp39-cp39-macosx_10_10_x86_64.whl` with 
[Pillow-9.4.0-cp39-cp39-macosx_10_10_x86_64.whl.zip](https://github.com/python-pillow/Pillow/files/10336005/Pillow-9.4.0-cp39-cp39-macosx_10_10_x86_64.whl.zip)

**Comment by radarhere** [comment]
I've created https://github.com/python-pillow/pillow-wheels/pull/357 as a possible solution to this. It downgrades the dependency for macOS < 11 and unfortunately 64-bit macOS >= 11 Python < 3.10, but otherwise keeps the latest version.

**Comment by denis-bz** [comment]
@radarhere, thanks for taking the time to look at this; I should've recognized the problem (10.14) sooner, my fault.
It's I think a general problem for all macOS wheels:
* Apple makes macOS upgrades difficult-to-impossible (planned obsolescence)
* wheelbuilders can't test on old versions.

Any idea how other people e.g. conda handle this ?
A line in `METADATA` like `External-requires = macos 10.15` would be useful --
if anybody checked it ...

**Comment by radarhere** [comment]
The wheel you would have been downloading is Pillow-9.4.0-cp39-cp39-macosx_10_10_x86_64.whl. In my PR, I've found that the end of the wheel name can be macosx_11_0_x86_64.whl, meaning that the wheel would only be used on macOS >= 11, but from what I can see, that ending is only supported for Python 3.10 and 3.11.

This is making use of PEP 425. `python3 -m pip debug --verbose` will list other tags that can be used to specify what platforms a wheel will work on.

**Comment by radarhere** [comment]
https://github.com/python-pillow/pillow-wheels/pull/358 has been merged. Future Pillow wheels should work on macOS 10.14 again.

**Comment by radarhere** [comment]
I've uploaded new Pillow 9.4.0 wheels. Installation should now work.
